### PR TITLE
[Consensus voting] Handling of invalid votes

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/5848

### Context

This PR implements several features:
- Extends `hotstuff.Notifier` with new callback which can be used to report votes for invalid proposal
- Updated implementors of notifier interface
- Added handling of invalid votes into `VoteAggregatorV2`.

### CODE REVIEW

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go [hunk 1]
```diff
@@ -185,8 +185,25 @@ func (va *VoteAggregatorV2) AddBlock(block *model.Proposal) error {
 	return nil
 }
 
-func (va *VoteAggregatorV2) InvalidBlock(block *model.Proposal) error {
-	panic("implement me")
+// InvalidBlock notifies the VoteAggregator about an invalid proposal, so that it can process votes for the invalid
+// block and slash the voters.
```
Comment (AlexHentschel) [inline consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:189]: lets document the expected error returns:

```suggestion
// block and slash the voters.
// Expected error returns during normal operations:
//  * mempool.DecreasingPruningHeightError if proposal's view has already been pruned 
```

could you also add the same comment to the interface, please https://github.com/onflow/flow-go/blob/e371dc223162a1f7496980345ecfae507d88023a/consensus/hotstuff/vote_aggregator_v2.go#L50

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go [hunk 2]
```diff
@@ -185,8 +185,25 @@ func (va *VoteAggregatorV2) AddBlock(block *model.Proposal) error {
...
+		if proposal.Block.BlockID == vote.BlockID {
+			va.notifier.OnVoteForInvalidBlockDetected(vote, proposal)
+		}
+	}
+
+	block := proposal.Block
+	collector, _, err := va.collectors.GetOrCreateCollector(block.View)
+	if err != nil {
+		return fmt.Errorf("could not retrieve collector for invalid proposal (%x) at view %d: %w",
+			block.BlockID, block.View, err)
```
Comment (AlexHentschel) [inline consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:201]: `VoteCollector` collects votes for a specific view (not a block)

```suggestion
		return fmt.Errorf("could not retrieve vote collector for view %d: %w", block.View, err)
```

File: consensus/hotstuff/votecollector/statemachine.go [hunk 3]
```diff
@@ -190,6 +190,11 @@ func (m *VoteCollector) ProcessBlock(proposal *model.Proposal) error {
 	}
 }
 
+// RegisterVoteConsumer effectively delegates call to votes cache
```
Comment (AlexHentschel) [inline consensus/hotstuff/votecollector/statemachine.go:193]: lets add the full goDoc here to retain all the details from the interface specification:

```suggestion
// RegisterVoteConsumer registers a VoteConsumer. Upon registration, the collector
// feeds all cached votes into the consumer in the order they arrived.
// CAUTION, VoteConsumer implementations must be
//  * NON-BLOCKING and consume the votes without noteworthy delay, and
//  * CONCURRENCY SAFE
```

**Review by AlexHentschel — APPROVED** [review]
Very nice     🎸  🎶🎵 🎤 

I found your terminology here very accurate: https://github.com/onflow/flow-go/blob/e371dc223162a1f7496980345ecfae507d88023a/consensus/hotstuff/notifications/slashing_violation_consumer.go#L42 
Unfortunately, the corresponding (already existing) nomenclature in [`LogConsumer`](https://github.com/onflow/flow-go/blob/e371dc223162a1f7496980345ecfae507d88023a/consensus/hotstuff/notifications/log_consumer.go) is inaccurate: https://github.com/onflow/flow-go/blob/e371dc223162a1f7496980345ecfae507d88023a/consensus/hotstuff/notifications/log_consumer.go#L133
* It would be great if you could replace _all occurrences_ of `vote_id` in `LogConsumer` with `voted_block_id`

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go [hunk 4]
```diff
@@ -185,8 +185,26 @@ func (va *VoteAggregatorV2) AddBlock(block *model.Proposal) error {
...
+			va.notifier.OnVoteForInvalidBlockDetected(vote, proposal)
+		}
+	}
+
+	block := proposal.Block
+	collector, _, err := va.collectors.GetOrCreateCollector(block.View)
+	if err != nil {
+		return fmt.Errorf("could not retrieve vote collector for view %d: %w", block.View, err)
+	}
+	// registering vote consumer will deliver all previously cached votes in strict order
```
Comment (zhangchiqing) [inline consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:204]: How does it handle the previously cached votes? 

If there are votings received before receiving the invalid block, we will also mark those votes as invalid, right?

Comment (durkmurder) [reply consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:204]: Yeap, when registering consumer it will deliver all previously cached votes

**Review by zhangchiqing — APPROVED** [review]
Nice work!

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors merge

**Pull request merged**